### PR TITLE
Improvements to handling and indication of recoverable errors

### PR DESCRIPTION
**Pull request by justyn**
These commits make three separate changes relating to the handling of recoverable exceptions.

### Background
The imap module treats some errors as recoverable rather than fatal. When these errors occur, the module will keep trying to check mail. Currently there is no indication that there has been a problem retrieving the mail count, and the module continues showing the last count.

### Changes in this PR
1) Maintain a list of recoverable exceptions so that it can be adjusted according to how the module is configured
2) Add a recoverable exception that is only thrown when using OAuth
3) Color the module output as *degraded* when a recoverable error means that the mail count is no longer necessarily up to date

### CODE REVIEW

**Comment by justyn** [comment]
I've just noticed PR #1881 which will conflict with this one. Probably best work on merging that first then I'll update this.

**Comment by girst** [comment]
I apologize for the negativity, but it is not a good idea to show recoverable errors to the user: many of them are even expected (like a logout-timeout) and then immediately retried (that's why they are even only visible in the console when debug is enabled--they don't provide valuable info to the user, only to the developer). I've been collecting erros, so here's a sample, so you can see what kind of info is in them:

```
2019-01-11 13:35:19 WARNING Module `imap 3`: Recoverable error - The read operation timed out
2019-01-11 13:38:26 WARNING Module `imap 3`: Recoverable error - command: CAPABILITY => Disconnected for inactivity.
```

If that happens, we immediately try again, and that (in 99% of cases) fixes the error--no need to tell the user. many imap servers are held together by duct tape and wishful thinking: them throwing errors is par for the course and the best we can do is to try again and be as fault-tolerant as possiblei

Adding TransportError to the recoverable errors is a good idea. Can you change the PR to just do the `tuple(self.recoverable_errors)`-stuff?

File: py3status/modules/imap.py [hunk 1]
```diff
@@ -177,8 +192,12 @@ def _check_if_idle(self, connection):
     def _get_creds(self):
         from google_auth_oauthlib.flow import InstalledAppFlow
         from google.auth.transport.requests import Request
+        from google.auth.exceptions import TransportError
```
Comment (girst) [inline py3status/modules/imap.py]: changes from here on to the end are good -- like it :-)

**Comment by girst** [comment]
also, the difference between a recoverable error and a fatal error is just whether we need to back off a bit. here are some fatal errors I've collected and what they mean:

```
2018-02-17 16:03:14 ERROR Module `imap 1`: Fatal error - b'[AUTHENTICATIONFAILED] LOGIN Invalid credentials'
2018-02-17 16:03:19 ERROR Module `imap 1`: Fatal error - b'[UNAVAILABLE] LOGIN Server error - Please try again later'
        those two are from yahoo. this does not mean that the password is wrong, it is a server bug: tearing down the connection and trying again works 100% of the time.
2018-03-19 17:25:09 ERROR Module `imap 0`: Fatal error - command SEARCH illegal in state AUTH, only allowed in states SELECTED
2019-03-10 00:44:05 ERROR Module `imap 0`: Fatal error - command SELECT illegal in state NONAUTH, only allowed in states AUTH, SELECTED
        we have been force-logged-out. retrying in a minute works all of the time. 
2020-01-14 12:15:42 ERROR Module `imap 0`: Fatal error - b'[SERVERBUG] LOGIN Server error - Please try again later'
        well, can't do much here, but try again. 
```

File: py3status/modules/imap.py [hunk 2]
```diff
@@ -129,7 +134,7 @@ def post_config_hook(self):
             raise ValueError("Unknown security protocol")
 
     def imap(self):
-        # I -- acquire mail_count
+        # acquire mail_count and start thread if using IDLE
```
[the remark this answers is not in the document]
Comment (justyn) [reply py3status/modules/imap.py]: Sure, that's partly down to me misunderstanding how the IDLE thread was working

**Comment by justyn** [comment]
Okay no problem, perhaps with some discussion we can find another way around the problem.

The issue I have is that if there is a network issue and the module is not actually updating, I am unaware of that and think I don't have mail, when in fact I do.

I frequently use public wifi, particularly on trains. So it is common to be in a situation that the connection is both unreliable and also involving captive portals.

It seems reasonable to me to colour the output to indicate that it isn't actually up to date, but I understand that isn't for everyone. Although technically this is configurable by setting the `degraded` colour for the module to match `good`, of course that isn't the default setup.

I could make this notification of recoverable errors an option, and off by default?

File: py3status/modules/imap.py [hunk 3]
```diff
@@ -115,11 +115,16 @@ def post_config_hook(self):
         # class variables:
         self.mail_count = NO_DATA_YET
         self.connection = None
-        self.mail_error = None  # cannot throw self.py3.error from thread
+        self.mail_error = ""  # cannot throw self.py3.error from thread
```
[the remark this answers is not in the document]
Comment (justyn) [reply py3status/modules/imap.py]: This was simply to allow distinguishing between fatal and recoverable errors.

**Comment by girst** [comment]
On Fri, Jan 17, 2020 at 10:42:44AM -0800, Justyn Butler wrote:
>The issue I have is that if there is a network issue and the module is 
>not actually updating, I am unaware of that and think I don't have 
>mail, when in fact I do.
>
>I frequently use public wifi, particularly on trains. So it is common 
>to be in a situation that the connection is both unreliable and also 
>involving captive portals.

I have this use case as well. this is actually a limitation of TCP 
sockets: they can't tell you when the underlying network breaks. what 
you have to do is to set cache_timeout to a lower value (IDLE will be 
cancelled and re-established every $cache_timeout seconds).

>It seems reasonable to me to colour the output to indicate that it 
>isn't actually up to date, but I understand that isn't for everyone.  

I'm actually not sure this would help anything, since we'll keep idling 
until cache_timeout is reached, at which point we'd log the error and 
immediately try again, which will clear the error.

>I could make this notification of recoverable errors an option, and off 
>by default?

again, i don't think it has the desired effect. with the retry logic in 
place, there should hardly be any of those problems left.

File: py3status/modules/imap.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (justyn) [reply py3status/modules/imap.py]: ...so discussion in the main comments may influence this.

[the remark this answers is not in the document]
Comment (girst) [reply py3status/modules/imap.py]: yep. for the record: for the end user, the distinction between fatal and recoverable errors is moot. this is just interesting for the developer to back off a bit.

**Comment by justyn** [comment]
Sorry, I'm not sure I follow, or perhaps I didn't explain well.

Given this example scenario that the wifi drops out or stops functioning for any reason...

**With the module as is**: the value for {unseen} remains fixed at the last value, and there is nothing in my i3bar to indicate that in fact it isn't up to date.

**With this imap-exceptions branch**: the value is still fixed at the last value, but the colour of the value in the i3bar changes (subtly, because I have chosen the value of `degraded` to be only a little different to `good` to avoid distraction).

I appreciate that the new retry logic could help the module stay connected in some situations, but I'm talking about the scenario that the network is unreachable for some reason (underlying mobile/cellular data has dropped out, or the network is redirecting to a captive portal, etc).

**Comment by girst** [comment]
i see what you're getting at now. yes, the old value will stay until a new value can be computed.


the following is long and rambly, apologies!

What I'm having trouble understanding is how the degraded color will help: it has been a while since I wrote this code, but my understanding is that once py3status figures out that we should display `degraded` (assuming your patch was applied), we are already retrying to fix it. 

here's how i picture it:

when our imap connection dies, IDLE will still just wait for a (never-coming) answer from the server. some time later, we will reach the timeout, enter the exception handling path (setting mail-error), and then retry until it works again. when you then notice the degraded state, I presume you restart i3(status), and that "fixes it"?

would you mind enabling `debug = True` in your imap config and keeping an eye on what errors are raised specifically when you hop between networks? if possible, with the retry logic in place. I'd like to figure out exactly what's wrong here.

the trouble is, "fatal" errors are mostly uninteresting to users. the only exception is, when our network connection disappears, and py3status stops updating. adding color in this case is suboptimal--i'd rather make a fix that will automatically re-restablish the connection. and detecting when we lose network connections is very hard with TCP, as the socket itself can't distinguish packet loss (expected) from a severed connection (the problem in question). best we can do is the timeout (i'm starting to repeat myself, sorry).

getting back to the color: once we made the decision to add `degraded`, we are already in the process of reestablishing the connection. since the computer might not be connected to a network any more, we don't try every second, but only every `cache_timeout` seconds. I think this is what you're critiquing: it takes too long. I think the default of 60sec is fine as a default, but for you you might want to set it lower. what do you have it set to now?

**Comment by girst** [comment]
**tl;dr:** what I'm getting at is that adding colors for all fatal erros will have way too many false-positives (which is why i haven't implemented it). this problem is better solved at the root--what to do when our connection disappears. the best I've come up with is decreasing `cache_timeout`

**Comment by justyn** [comment]
> when our imap connection dies, IDLE will still just wait for a (never-coming) answer from the server. some time later, we will reach the timeout, enter the exception handling path (setting mail-error), and then retry until it works again. when you then notice the degraded state, I presume you restart i3(status), and that "fixes it"?
> 
> would you mind enabling `debug = True` in your imap config and keeping an eye on what errors are raised specifically when you hop between networks? if possible, with the retry logic in place. I'd like to figure out exactly what's wrong here.

Actually no, I don't need to restart i3. The value seems to stay in the degraded colour until the network comes back.

As it happens I wrote the previous comments on a train and I have an example from my log:

```
2020-01-17 18:51:44 INFO method imap returned {'color': '#FFE5CC', 'composite': [{'full_text': 'gmail: 0', 'markup': 'pango', 'min_width': 80, 'align': 'center', 'separator': True, 'separator_block_width': 30, 'instance': 'gmail 0', 'name': 'imap', 'color': '#FFE5CC'}], 'instance': 'gmail', 'name': 'imap'} 
2020-01-17 18:51:44 WARNING Module `imap gmail`: Recoverable error - [Errno -3] Temporary failure in name resolution
2020-01-17 18:51:44 INFO clearing cache for method imap
```


> the trouble is, "fatal" errors are mostly uninteresting to users. the only exception is, when our network connection disappears, and py3status stops updating. adding color in this case is suboptimal--i'd rather make a fix that will automatically re-restablish the connection. and detecting when we lose network connections is very hard with TCP, as the socket itself can't distinguish packet loss (expected) from a severed connection (the problem in question). best we can do is the timeout (i'm starting to repeat myself, sorry).
> 
> getting back to the color: once we made the decision to add `degraded`, we are already in the process of reestablishing the connection. since the computer might not be connected to a network any more, we don't try every second, but only every `cache_timeout` seconds. I think this is what you're critiquing: it takes too long. I think the default of 60sec is fine as a default, but for you you might want to set it lower. what do you have it set to now?

I have cache_timeout set to the default, but I don't _think_ I'm critiquing how long this takes. I may not be fully up to speed yet, sorry!

Also above you say "fatal" errors, whereas I think I'm talking about recoverable ones, and that's what the patch is looking for to set degraded.

However I do take your point that concentrating on what kind of error occurs (fatal, recoverable) is a red herring to the user.

How about instead we ignore errors and use a retry counter? After _this many_ unsuccessful attempts to connect (default 5 or something, set in the config) colour the output degraded until we successfully make a connection.

**Comment by justyn** [comment]
Another kind of recoverable error I frequently have, this one caused by a captive portal interfering (and from the same train journey as the other error!):

```
2020-01-17 18:27:33 WARNING Module `imap gmail`: Recoverable error - HTTPSConnectionPool(host='oauth2.googleapis.com', port=443): Max retries exceeded with url: /token (Caused by SSLError(SSLCertVerificationError("hostname 'oauth2.googleapis.com' doesn't match 'www.ombord.info'")))
```

**Comment by girst** [comment]
On Fri, Jan 17, 2020 at 11:44:25AM -0800, Justyn Butler wrote:
>Actually no, I don't need to restart i3. The value seems to stay in the 
>degraded colour until the network comes back.

oh, ok. wrong assumption from my part.

>2020-01-17 18:51:44 WARNING Module `imap gmail`: Recoverable error - [Errno -3] Temporary failure in name resolution

the dreadful socket.error. however, socket.errors are also raised 
everytime IDLE runs in the expected timeout, so not universally an 
error.

>I have cache_timeout set to the default, but I don't _think_ I'm 
>critiquing how long this takes. I may not be fully up to speed yet, 
>sorry!

i was assuming you took some action when the color changed to degraded, 
but now i believe you "just want to know" in such a case.

>How about instead we ignore errors and use a retry counter? After _this 
>many_ unsuccessful attempts to connect (default 5 or something, set in 
>the config) colour the output degraded until we successfully make a 
>connection.

that might be a sensible way to move forward, but that would have to 
wait until the retry-PR is merged. maybe set this PR to work-in-progress 
and we can pick it up together then?

**Comment by girst** [comment]
On Fri, Jan 17, 2020 at 11:47:47AM -0800, Justyn Butler wrote:
>2020-01-17 18:27:33 WARNING Module `imap gmail`: Recoverable error - HTTPSConnectionPool(host='oauth2.googleapis.com', port=443): Max retries exceeded with url: /token (Caused by SSLError(SSLCertVerificationError("hostname 'oauth2.googleapis.com' doesn't match 'www.ombord.info'")))

hmmm, can you get me a traceback for this one? in the error handler, put 
this somewhere:

import traceback
self.py3.log(traceback.format_exc())

i'm assuming this comes from the oauth login stuff and is a 
very-special-case, as we need to back off to not trigger the rate 
limiting. i propose handling this within the oauth-login-function.

**Comment by justyn** [comment]
> On Fri, Jan 17, 2020 at 11:44:25AM -0800, Justyn Butler wrote: Actually no, I don't need to restart i3. The value seems to stay in the degraded colour until the network comes back.
> oh, ok. wrong assumption from my part.

It sounds like the kind of exceptions we run into in our own real-world uses are slightly different, since in most cases I'm talking about situations where the inability to get the mail count is out of the module's control due to network issues.

> 2020-01-17 18:51:44 WARNING Module `imap gmail`: Recoverable error - [Errno -3] Temporary failure in name resolution
> the dreadful socket.error. however, socket.errors are also raised everytime IDLE runs in the expected timeout, so not universally an error.

I definitely need to understand better exactly what is happening when these exceptions are raised, because I'm not seeing "degraded" output during normal use.

> I have cache_timeout set to the default, but I don't _think_ I'm critiquing how long this takes. I may not be fully up to speed yet, sorry!
> i was assuming you took some action when the color changed to degraded, but now i believe you "just want to know" in such a case.
> How about instead we ignore errors and use a retry counter? After _this many_ unsuccessful attempts to connect (default 5 or something, set in the config) colour the output degraded until we successfully make a connection.
> that might be a sensible way to move forward, but that would have to wait until the retry-PR is merged. maybe set this PR to work-in-progress and we can pick it up together then?

Absolutely, that makes sense, thanks.

> hmmm, can you get me a traceback for this one? in the error handler, put this somewhere: import traceback self.py3.log(traceback.format_exc()) i'm assuming this comes from the oauth login stuff and is a very-special-case, as we need to back off to not trigger the rate limiting. i propose handling this within the oauth-login-function.

Will do.

**Comment by ultrabug** [comment]
Thanks for this constructive conversation mates.

What @justyn is trying to do here is actually a feature that I had in mind to normalize for all modules for py3status version 4 because the numerous modules depending on network do have a different behavior when facing a network outage.

If I'm understanding correctly what we're trying to achieve is this : "indicate that the module is stale and waiting for something to recover and update its output on the bar"

My first thought was to use the degraded color as well to indicate that, or place parenthesis around the output or a * ... never could make my mind on this, but the idea is there.

What I wanted to do is to tag modules with features so that a module tagged "network" would default to this kind of behavior.

But since that's for v4 and the road is long to there, I think we should get this done here as a nick mockup!

**Comment by girst** [comment]
On Sat, Jan 18, 2020 at 01:32:01PM -0800, Ultrabug wrote:
>If I'm understanding correctly what we're trying to achieve is this : 
>"indicate that the module is stale and waiting for something to recover 
>and update its output on the bar"

yes.

>My first thought was to use the degraded color as well to indicate 
>that, or place parenthesis around the output or a * ... never could 
>make my mind on this, but the idea is there.

if this is implemented globally, why not make it configurable? quick 
mock-up:

     global {
	on_network_outage = "format"
	outage_format = "{text}*"  # append an asterisk on a network outage
	# on_network_outage = "color_degraded"
	# on_network_outage = "" # or something else falsey to indicate no change
     }

>But since that's for v4 and the road is long to there, I think we 
>should get this done here as a nick mockup!

"nick mockup"?

**Comment by ultrabug** [comment]
> if this is implemented globally, why not make it configurable? quick
sure, good point!

> "nick mockup"?
edited; typo for "nice" :)

**Comment by girst** [comment]
On Sat, Jan 18, 2020 at 01:42:04PM -0800, Ultrabug wrote:
>> "nick mockup"?
>edited; typo for "nice" :)

duh; obvious in hindsight :/

thinking about this some more, network problems are complex. i believe 
that each module should inform py3status' core itself when it has lost 
connection. not only can the server we're speaking to go offline (while 
we're still online), but some services might be reachable without a 
connection to the internet (localhost, LAN). also, when the user's 
router loses connectivity to the internet, but the user's computer is 
still connected to the router. what about servers that are only 
available in an intranet? all this is hard (if not impossible) to detect 
without looking at the socket (i.e. by getting notified from the 
module).

would it make sense to extend py3.error() or py3.update() or add a new 
py3.network_outage() that allows networked modules to signal their 
staleness?

**Comment by justyn** [comment]
That's interesting, sounds like a good idea to have something standard for
network related modules.

My main request is that the indication should be able to be very subtle (if
desired).

As previously explained, network issues are a common problem for me. But my
main goal with using py3status has been to get the right balance between
information and distraction. It's very important to me that a status bar
doesn't disrupt me.

That's why personally I really like the option of a colour...it is possible
to set this to be a subtle change, and therefore not actually distract
attention...I want to know that the information is stale when I look but
without my eye actually being drawn away from my work for something as
minor as a network issue.

**Comment by ultrabug** [comment]
Well as I said, this is a v4 topic so I think that if this feature is not default and can help imap users then your contribution could get in @justyn 

There are conflicts, could you plz take care of them?

**Comment by girst** [comment]
I've had a go at implementing the basic version with just setting the color to degraded (disable-able in the config). I wasn't happy with modifying `mail_error` and @justyn's patch introduces a bug regarding setting colors, so I implemented this on top of master.

@justyn: if we were to go with my implementation, we'd still need the oauth error handling. i suggest splitting this off into a second pull request, so we can discuss how we want to handle oauth errors in detail.

**Comment by justyn** [comment]
> I've had a go at implementing the basic version with just setting the color to degraded (disable-able in the config). I wasn't happy with modifying `mail_error` and @justyn's patch introduces a bug regarding setting colors, so I implemented this on top of master.

Great, I'll try it out.

> @justyn: if we were to go with my implementation, we'd still need the oauth error handling. i suggest splitting this off into a second pull request, so we can discuss how we want to handle oauth errors in detail.

Sure, I'll create a separate PR to keep things focused.

**Comment by justyn** [comment]
Actually @girst just to check we are on the same page.

The new PR will really just be 2 of the 3 commits from this one, updated to run on top of master.

We can discuss/test from that point, and I realise that you are thinking about special handling of oauth errors, but right now should it not just be treated the same as one of the other network errors?

**Comment by girst** [comment]
On Sun, Jan 19, 2020 at 06:57:20AM -0800, Justyn Butler wrote:
>The new PR will really just be 2 of the 3 commits from this one, 
>updated to run on top of master.

if you mean "list recoverable errors as attribute of instance.", i'd 
consider this part of the oauth error handling changes.

>We can discuss/test from that point, and I realise that you are 
>thinking about special handling of oauth errors, but right now should 
>it not just be treated the same as one of the other network errors?

no, i think oauth errors should be handled differently than network 
errors and more like login errors (part of fatal errors). let's discuss 
the details on the other PR ;-)

**Comment by justyn** [comment]
@girst This PR is technically redundant and I was going to close, but actually there is something I'd like to ask about the handling of network errors. It is more related to the handling of errors than marking output as stale, or OAuth, so perhaps here is the best place.

At the moment when the connection is lost the imap module tries to reconnect 3 times in quick succession every read_timeout. I got the impression this triple-try is something to do with misconfigured IMAP servers?

If so, is it possible to set the retry value to 1, instead of 3, for simple network errors?

**Comment by girst** [comment]
On Sun, Jan 26, 2020 at 05:50:11AM -0800, Justyn Butler wrote:
>At the moment when the connection is lost the imap module tries to 
>reconnect 3 times in quick succession every read_timeout. I got the 
>impression this triple-try is something to do with misconfigured IMAP 
>servers?

correct. mostly its servers that log you out or disconnect after a 
period of not sending a command (which you can't during idle), but I've 
seen other errors as well, all fixed by retrying.

the number of retries was chosen arbitrarily, in case you wonder.

>If so, is it possible to set the retry value to 1, instead of 3, for 
>simple network errors?

depends: do we gain anything? (honest question) if not, i'd rather not 
complicate the code even more, since its a lot to read through already, 
especially since the new formatting.

**Comment by justyn** [comment]
> depends: do we gain anything? (honest question) if not, i'd rather not complicate the code even more, since its a lot to read through already, especially since the new formatting.

Well the small gain would be reducing the extra attempts to connect in situations where it is guaranteed to be fruitless, but I certainly take your point about further complicating the code.

I'll close this PR now.
